core/src/main/java/software/amazon/awssdk/core/auth/HttpCredentialsProvider.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright 2010-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.core.auth;
+
+import com.fasterxml.jackson.databind.JsonMappingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import java.io.IOException;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Optional;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.SdkClientException;
+import software.amazon.awssdk.core.internal.CredentialsEndpointProvider;
+import software.amazon.awssdk.core.internal.HttpCredentialsUtils;
+import software.amazon.awssdk.core.util.ComparableUtils;
+import software.amazon.awssdk.core.util.DateUtils;
+import software.amazon.awssdk.core.util.json.JacksonUtils;
+import software.amazon.awssdk.utils.SdkAutoCloseable;
+import software.amazon.awssdk.utils.Validate;
+import software.amazon.awssdk.utils.cache.CachedSupplier;
+import software.amazon.awssdk.utils.cache.NonBlocking;
+import software.amazon.awssdk.utils.cache.RefreshResult;
+
+/**
+ * Helper class that contains the common behavior of the CredentialsProviders that loads the credentials from a local endpoint on
+ * a container (e.g. an EC2 instance).
+ */
+@SdkInternalApi
+abstract class HttpCredentialsProvider implements AwsCredentialsProvider, SdkAutoCloseable {
+    private final CachedSupplier<AwsCredentials> credentialsCache;
+
+    HttpCredentialsProvider(Builder<?, ?> builder) {
+        this(builder.asyncCredentialUpdateEnabled, builder.asyncThreadName);
+    }
+
+    HttpCredentialsProvider(Boolean asyncCredentialUpdateEnabled, String asyncThreadName) {
```

Comment:
can we use `boolean` for `asyncCredentialUpdateEnabled`? we would never need to worry about NPE for it.

Reply:
I can't think of a good reason why not